varname/utils.py

```diff
@@ -225,3 +230,75 @@ def debug_ignore_frame(
         msg = (f'{msg} [In {frameinfo.function!r} at '
                f'{frameinfo.filename}:{frameinfo.lineno}]')
     sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+def funcsig_from_ast_call(node: ast.Call,
+                          globs: dict,
+                          locs: dict) -> inspect.Signature:
+    """Get the function signature from an ast.Call node"""
+    expr = ast.Expression(node.func)
+    code = compile(expr, '<ast-call>', 'eval')
+    func = eval(code, globs, locs) # pylint: disable=eval-used
```

Comment:
This is a bit risky, the function expression could be anything. Options:

1. Clearly document that this happens in `argname`
2. Only allow variables as the function
3. https://github.com/alexmojaki/pure_eval, although this will only some basic expressions

Reply:
This is not the optimal solution to get the function, how about this one:
```python
def getfunc():
    frame = sys._getframe(1)
    finfo = inspect.getframeinfo(frame)
    funcname = finfo.function
    return frame.f_locals.get(funcname, frame.f_globals.get(funcname))
    
def func():
    print(getfunc())
    
func() # <function func at 0x7f6d685e8170>
```

Reply:
That's sort of option 2 - you're saying the function must exist as a variable, although technically it may not be called that way. But `finfo.function` is just `frame.f_code.co_name` and this can go wrong in several ways:

1. If the function is decorated, the variable name and code name probably won't match
2. If the function was imported with `import ... as`
3. If the call is a method `self.foo(bar)` and the method is not accessible as a variable but there also happens to be a  different function `foo` in local/global variables
4. If the function is defined as a lambda

Instead `funcname` should come from the AST.

Reply:
1. It is a bit complicated for decorated functions, as we have to define whose arguments they are or to be targeted, either the wrapper's or the original function's. And it may not make sense either way. To do it for the original function's arguments, which is, probably what the users want, then one has to define the wrapper with the same signature. But we can not assume that on our end. While to do it for the wrapper's arguments? Probably no one cares. People may just throw `wrapper(*args, **kwargs)` in there.
2. It should work for this case. Because the function will exist as a variable in the namespace, which can be fetched by `frame.f_locals` or `frame.f_globals`
3. True, it is not perfect either.
4. It works if `lambda` is passed to a variable, but definitely not like this: `(lambda x: argname(x))(1)`

The other option I am thinking of is to ask the user to pass the function to `argname` explicitly, and only apply the possible strategies above to fetch the function if they don't.

Reply:
I'm not sure you understand. I'm saying the name of the function used to look up the variable should come from the AST, not the code name which is `finfo.function`.

1. You're right that we probably want the original arguments and not the wrapper, so I've deleted my other suggestion. But either way if the code name is `wrapper` that's probably not the name of the variable being used.
2. No it shouldn't, if I write `from mod import foo as bar` then `foo` is the code name but the user calls `bar()` and only `bar` is in the locals/globals.
3. It's not just not perfect, it could lead to silent wrong behaviour.
4. No it doesn't, e.g. `func = lambda: print(getfunc()); func()` doesn't work.

I think your last idea is pretty good but it's not clear when they should pass the function and when it's not really worth it.

Reply:
Oh, now I see your point. My bad. All my replies were not based on 

> Instead `funcname` should come from the AST.

which I agree. Then using `frame.f_locals/f_globals` is not reliable at all. Combined with your suggestions:

1. Guide and encourage the users to pass the function explicitly. Actually, they should always do so. Explain in the doc that the function will be fetched using `eval`, which will be risky, or using `pure_eval`, which is limited and `pure_eval` has to be installed.
2. In case they don't, follow your suggestions (limiting node types of `node.func` or using `pure_eval`). If failed, raise an error to pass the function explicitly.

Reply:
I'll clarify. The options are (not in the same order as before):

1. Only allow `node.func` to be an `ast.Name`. In that case locals/globals is fine, but only with `funcname = node.func.id`.
2. Only allow `node.func` to be a node that pure_eval can evaluate. Currently this is restricted to Name, Attribute, and Subscript, and combinations of those and literals. But a node of that type might still be refused by pure_eval as unsafe.
3. Allow `node.func` to be anything, use eval, but document it loudly.
4. Combine 1 and 2: if it's not a Name, try pure_eval, raising an error if it's not installed. If pure_eval is a strict requirement of the library, no point in doing this.
5. Combine all 3: if it's not a Name and pure_eval is either not installed or fails, use eval() but raise a warning or only allow this if they passed a flag to argname.

Do not limit the types of node.func to anything more than Name and then try to evaluate that yourself without pure_eval, that's literally what pure_eval is for. Even an innocent looking attribute access can trigger arbitrary code.

Reply:
Clear enough.

For the decorated functions, other than the first option, looks like we can't retrieve the signature of the original function by `.__wrapped__`. And even for the first one, the wrapper has to be decorated with `functools.wraps`.

Maybe we have to bring similar stuff from `ignore` to support decorated functions, that is to get the original function from the frame where it is called.